File: cpp/src/tum.cpp
#include "rgbd/tum.hpp"

#include <Eigen/Geometry>

#include <algorithm>
#include <cmath>
#include <fstream>
#include <iomanip>
#include <sstream>
#include <stdexcept>
#include <tuple>
#include <utility>

namespace rgbd {
namespace {

using FileRow = std::pair<double, std::filesystem::path>;
using PoseRow = std::pair<double, Mat4>;

bool skip_line(const std::string& line) {
    const auto first = line.find_first_not_of(" \t\r\n");
    return first == std::string::npos || line[first] == '#';
}

std::vector<FileRow> read_index(const std::filesystem::path& path) {
    std::ifstream stream(path);
    if (!stream) {
        throw std::runtime_error("cannot open TUM index: " + path.string());
    }
    std::vector<FileRow> rows;
    std::string line;
    while (std::getline(stream, line)) {
        if (skip_line(line)) {
            continue;
        }
        std::istringstream values(line);
        double timestamp = 0.0;
        std::string filename;
        if (!(values >> timestamp >> filename)) {
            throw std::runtime_error("invalid TUM index row in " + path.string());
        }
        rows.emplace_back(timestamp, path.parent_path() / filename);
    }
    return rows;
}

std::vector<PoseRow> read_ground_truth(const std::filesystem::path& path) {
    std::ifstream stream(path);
    if (!stream) {
        throw std::runtime_error("cannot open TUM ground truth: " + path.string());
    }
    std::vector<PoseRow> rows;
    std::string line;
    while (std::getline(stream, line)) {
        if (skip_line(line)) {
            continue;
        }
        std::istringstream values(line);
        double timestamp = 0.0;
        double tx = 0.0;
        double ty = 0.0;
        double tz = 0.0;
        double qx = 0.0;
        double qy = 0.0;
        double qz = 0.0;
        double qw = 0.0;
        if (!(values >> timestamp >> tx >> ty >> tz >> qx >> qy >> qz >> qw)) {
            throw std::runtime_error("invalid TUM ground-truth row in " + path.string());
        }
        const Eigen::Quaterniond quaternion(qw, qx, qy, qz);
        Mat4 pose = Mat4::Identity();
        pose.block<3, 3>(0, 0) = quaternion.normalized().toRotationMatrix();
        pose.block<3, 1>(0, 3) = Vec3(tx, ty, tz);
        rows.emplace_back(timestamp, pose);
    }
    return rows;
}

std::vector<std::pair<std::size_t, std::size_t>> associate_one_to_one(
    const std::vector<double>& first_times,
    const std::vector<double>& second_times,
    double max_offset_s) {
    using Candidate = std::tuple<double, std::size_t, std::size_t>;
    std::vector<Candidate> candidates;
    for (std::size_t first = 0; first < first_times.size(); ++first) {
        const auto left = std::lower_bound(second_times.begin(), second_times.end(), first_times[first] - max_offset_s);
        const auto right = std::upper_bound(second_times.begin(), second_times.end(), first_times[first] + max_offset_s);
        for (auto position = left; position != right; ++position) {
            const std::size_t second = static_cast<std::size_t>(position - second_times.begin());
            candidates.emplace_back(std::abs(first_times[first] - *position), first, second);
        }
    }
    std::sort(candidates.begin(), candidates.end());
    std::vector<bool> used_first(first_times.size(), false);
    std::vector<bool> used_second(second_times.size(), false);
    std::vector<std::pair<std::size_t, std::size_t>> matches;
    for (const auto& [offset, first, second] : candidates) {
        static_cast<void>(offset);
        if (!used_first[first] && !used_second[second]) {
            used_first[first] = true;
            used_second[second] = true;
            matches.emplace_back(first, second);
        }
    }
    std::sort(matches.begin(), matches.end());
    return matches;
}

}  // namespace

std::vector<RgbdFrame> load_tum_rgbd_frames(
    const std::filesystem::path& dataset_dir,
    double max_depth_time_offset_s,
    double max_ground_truth_time_offset_s) {
    const std::vector<FileRow> rgb = read_index(dataset_dir / "rgb.txt");
    const std::vector<FileRow> depth = read_index(dataset_dir / "depth.txt");
    const std::vector<PoseRow> ground_truth = read_ground_truth(dataset_dir / "groundtruth.txt");
    if (rgb.empty() || depth.empty() || ground_truth.empty()) {
        throw std::runtime_error("RGB, depth, and ground-truth indices must be non-empty");
    }
    std::vector<double> rgb_times;
    std::vector<double> depth_times;
    std::vector<double> ground_truth_times;
    for (const auto& row : rgb) rgb_times.push_back(row.first);
    for (const auto& row : depth) depth_times.push_back(row.first);
    for (const auto& row : ground_truth) ground_truth_times.push_back(row.first);
    const auto rgb_depth_matches = associate_one_to_one(rgb_times, depth_times, max_depth_time_offset_s);
    std::vector<double> matched_rgb_times;
    for (const auto& [rgb_index, depth_index] : rgb_depth_matches) {
        static_cast<void>(depth_index);
        matched_rgb_times.push_back(rgb[rgb_index].first);
    }
    const auto rgbd_gt_matches = associate_one_to_one(
        matched_rgb_times, ground_truth_times, max_ground_truth_time_offset_s);

    std::vector<RgbdFrame> frames;
    frames.reserve(rgbd_gt_matches.size());
    for (const auto& [rgb_depth_index, gt_index] : rgbd_gt_matches) {
        const auto [rgb_index, depth_index] = rgb_depth_matches[rgb_depth_index];
        const auto& [rgb_time, rgb_path] = rgb[rgb_index];
        const double depth_offset = std::abs(depth[depth_index].first - rgb_time);
        const double gt_offset = std::abs(ground_truth[gt_index].first - rgb_time);
        frames.push_back({
            rgb_time,
            rgb_path,
            depth[depth_index].second,
            depth_offset,
            ground_truth[gt_index].second,
            gt_offset,
        });
    }
    if (frames.size() < 2) {
        throw std::runtime_error("fewer than two associated TUM RGB-D frames");
    }
    return frames;
}

std::string pose_to_tum_row(double timestamp, const Mat4& pose) {
    const Eigen::Quaterniond quaternion(pose.block<3, 3>(0, 0));
    const Vec3 translation = pose.block<3, 1>(0, 3);
    std::ostringstream row;
    row << std::fixed << std::setprecision(9)
        << timestamp << ' ' << translation.x() << ' ' << translation.y() << ' ' << translation.z() << ' '
        << quaternion.x() << ' ' << quaternion.y() << ' ' << quaternion.z() << ' ' << quaternion.w();
    return row.str();
}

}  // namespace rgbd

File: src/tum.py
"""TUM RGB-D text-file parsing and timestamp association utilities."""

from __future__ import annotations

from bisect import bisect_left, bisect_right
from dataclasses import dataclass
from pathlib import Path

import numpy as np
from scipy.spatial.transform import Rotation


@dataclass(frozen=True)
class RgbdPair:
    """One timestamp-associated RGB/depth pair, independent of evaluation data."""

    timestamp: float
    rgb_path: Path
    depth_path: Path
    depth_time_offset_s: float


@dataclass(frozen=True)
class RgbdFrame(RgbdPair):
    """An RGB-D pair with an evaluation-only ground-truth camera pose."""

    ground_truth: np.ndarray  # 4x4 camera-to-world matrix
    ground_truth_time_offset_s: float


def _read_index(path: Path) -> list[tuple[float, Path]]:
    rows: list[tuple[float, Path]] = []
    for line in path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        stamp, name = line.split(maxsplit=1)
        rows.append((float(stamp), path.parent / name))
    return sorted(rows, key=lambda row: row[0])


def _associate_one_to_one(first_times: list[float], second_times: list[float], max_offset_s: float) -> list[tuple[int, int]]:
    """Return deterministic minimum-offset greedy matches without sample reuse."""
    if not np.isfinite(max_offset_s) or max_offset_s < 0.0:
        raise ValueError("max_offset_s must be finite and non-negative")
    candidates: list[tuple[float, int, int]] = []
    for first_index, first_time in enumerate(first_times):
        left = bisect_left(second_times, first_time - max_offset_s)
        right = bisect_right(second_times, first_time + max_offset_s)
        candidates.extend(
            (abs(first_time - second_times[second_index]), first_index, second_index)
            for second_index in range(left, right)
        )
    used_first: set[int] = set()
    used_second: set[int] = set()
    matches: list[tuple[int, int]] = []
    for _, first_index, second_index in sorted(candidates):
        if first_index not in used_first and second_index not in used_second:
            used_first.add(first_index)
            used_second.add(second_index)
            matches.append((first_index, second_index))
    return sorted(matches)


def _read_ground_truth(path: Path) -> list[tuple[float, np.ndarray]]:
    poses: list[tuple[float, np.ndarray]] = []
    for line in path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        values = [float(value) for value in line.split()]
        if len(values) != 8:
            raise ValueError(f"Expected 8 columns in {path}: {line}")
        timestamp, tx, ty, tz, qx, qy, qz, qw = values
        pose = np.eye(4)
        pose[:3, :3] = Rotation.from_quat([qx, qy, qz, qw]).as_matrix()
        pose[:3, 3] = [tx, ty, tz]
        poses.append((timestamp, pose))
    return sorted(poses, key=lambda row: row[0])


def load_tum_rgbd_pairs(
    dataset_dir: str | Path,
    *,
    max_depth_time_offset_s: float = 0.02,
) -> list[RgbdPair]:
    """Load deterministic one-to-one RGB/depth pairs without requiring poses."""
    root = Path(dataset_dir)
    rgb = _read_index(root / "rgb.txt")
    depth = _read_index(root / "depth.txt")
    if not rgb or not depth:
        raise ValueError("RGB and depth index files must both contain samples")
    matches = _associate_one_to_one(
        [item[0] for item in rgb], [item[0] for item in depth], max_depth_time_offset_s
    )
    pairs = [
        RgbdPair(
            rgb[rgb_index][0],
            rgb[rgb_index][1],
            depth[depth_index][1],
            abs(depth[depth_index][0] - rgb[rgb_index][0]),
        )
        for rgb_index, depth_index in matches
    ]
    if len(pairs) < 2:
        raise RuntimeError("Fewer than two timestamp-associated RGB-D pairs were found")
    return pairs


def associate_ground_truth(
    pairs: list[RgbdPair],
    ground_truth_path: str | Path,
    *,
    max_ground_truth_time_offset_s: float = 0.02,
) -> list[RgbdFrame]:
    """Attach evaluation poses one-to-one without changing the RGB-D loader."""
    ground_truth = _read_ground_truth(Path(ground_truth_path))
    if not ground_truth:
        raise ValueError("Ground-truth file must contain poses")
    matches = _associate_one_to_one(
        [pair.timestamp for pair in pairs],
        [item[0] for item in ground_truth],
        max_ground_truth_time_offset_s,
    )
    frames = [
        RgbdFrame(
            pairs[pair_index].timestamp,
            pairs[pair_index].rgb_path,
            pairs[pair_index].depth_path,
            pairs[pair_index].depth_time_offset_s,
            ground_truth[truth_index][1],
            abs(ground_truth[truth_index][0] - pairs[pair_index].timestamp),
        )
        for pair_index, truth_index in matches
    ]
    if len(frames) < 2:
        raise RuntimeError("Fewer than two timestamp-associated evaluation frames were found")
    return frames


def load_tum_rgbd_frames(
    dataset_dir: str | Path,
    *,
    max_depth_time_offset_s: float = 0.02,
    max_ground_truth_time_offset_s: float = 0.02,
) -> list[RgbdFrame]:
    """Associate RGB, depth, and ground truth one-to-one by minimum time offset.

    Candidate pairs inside each gate are sorted by offset and greedily consumed,
    matching the TUM association script's no-reuse rule. Poses follow TUM
    convention: camera to world.
    """
    root = Path(dataset_dir)
    pairs = load_tum_rgbd_pairs(root, max_depth_time_offset_s=max_depth_time_offset_s)
    return associate_ground_truth(
        pairs,
        root / "groundtruth.txt",
        max_ground_truth_time_offset_s=max_ground_truth_time_offset_s,
    )


def pose_to_tum_row(timestamp: float, pose: np.ndarray) -> str:
    """Serialize a camera-to-world matrix in the standard TUM trajectory format."""
    pose = np.asarray(pose, dtype=float)
    quaternion = Rotation.from_matrix(pose[:3, :3]).as_quat()
    translation = pose[:3, 3]
    values = [timestamp, *translation, *quaternion]
    return " ".join(f"{value:.9f}" for value in values)


def read_tum_trajectory(path: str | Path) -> tuple[list[float], list[np.ndarray]]:
    """Read ``timestamp tx ty tz qx qy qz qw`` trajectory rows."""
    timestamps: list[float] = []
    poses: list[np.ndarray] = []
    for line_number, raw_line in enumerate(Path(path).read_text(encoding="utf-8").splitlines(), start=1):
        line = raw_line.strip()
        if not line or line.startswith("#"):
            continue
        values = [float(value) for value in line.split()]
        if len(values) != 8:
            raise ValueError(f"Expected 8 TUM trajectory columns at line {line_number}: {raw_line}")
        timestamp, tx, ty, tz, qx, qy, qz, qw = values
        pose = np.eye(4)
        pose[:3, :3] = Rotation.from_quat([qx, qy, qz, qw]).as_matrix()
        pose[:3, 3] = [tx, ty, tz]
        timestamps.append(timestamp)
        poses.append(pose)
    if not poses:
        raise ValueError(f"No poses found in {path}")
    return timestamps, poses
